Decision Tree Visualizer

Rgtemze opened the pull request:
I realized that there is no example to visualize tree structure. So, we have been working on tree visualizing especially decision trees. We contacted the community on gitter and we are advised to create PR.

pratapvardhan commented:
It appears `__pycache__` folder and `.pyc` files are commited. You might want to uncheck them.

bryevdv commented:
I would recommend a rebase/force push to clean them up without noise commits at this point.

bryevdv commented:
@Rgtemze I noticed you added an MIT license. In order to avoid any confusion and keep things as simple as possible, all the examples should be under the same license as the general license at the top of the repo, i.e. BSD 3-clause. If this is not feasible from you, then this example might be better of in a different place that we can simply link to in the documentation. 

Additionally, please take some time to study the style, naming, and coding conventions of all the other examples, and update this code to match as  closely as possible.

bryevdv commented:
There's also an error preventing the code quality test from running:
```
E       UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```
Guessing there is a unicode character in one of the `.py` files? If it's not utf-8 the file probably needs a  `coding` comment at the top tho I'd say it's best if all files can be utf-8

Rgtemze commented:
@bryevdv Thank you for your interest. 
As it is our first contribution, we may have done something wrong; we will do our best to fix those problems.
Besides, we have been trying to locate that character which creates decoding problem since the beginning of the contribution. Hopefully, we will resolve this. 
Sincerely.

pratapvardhan commented on examples/app/bokeh-decision-tree/Decision_Tree/ID3_Decision_Tree/generate_bokeh_data.py:
```diff
@@ -0,0 +1,141 @@
+    elif level > 1:
+        if node.parentPointer:
+            node.depth = node.parentPointer.depth - 1
+        return sum([get_width(child, level-1) for child in node.children])
```
You could use `sum(get_width(child, level-1) for child in node.children)` removing the square brackets.

pratapvardhan commented on examples/app/bokeh-decision-tree/Decision_Tree/ID3_Decision_Tree/id3_decision_tree.py:
```diff
@@ -0,0 +1,394 @@
+    if numbers_greater_than_zero == 1:
+        return True
+    else:
+        return False
```
This could perhaps be `return numbers_greater_than_zero == 1` no?

pratapvardhan commented on examples/app/bokeh-decision-tree/Decision_Tree/ID3_Decision_Tree/id3_decision_tree.py:
```diff
@@ -0,0 +1,394 @@
+    for i in range(len(distributed_list_on_class_attr)):
+        if distributed_list_on_class_attr[i] == max_occurrence:
+            max_indexes.append(i)
+
```
For returning indices of max value in the list. You could consider.

```
[i for i, v in enumerate(distributed_list_on_class_attr) if v == max_occurrence]
```

pratapvardhan commented on examples/app/bokeh-decision-tree/templates/index.html:
```diff
@@ -0,0 +1,52 @@
+        <a href="mailto:y.umutciftci@gmail.com">
+            <img src="bokeh-decision-tree/static/images/umut.jpeg" width="120">
+        </a>
+    </div>
```
Not sure, what these links and images are for?

Rgtemze replied:
These are images of ours containing links to our mails and linked-in profiles in the About section of the web page.

bryevdv replied:
I don't think this kind of promotion is appropriate for examples in the repo, there are no other examples that do this. If you'd like to have an example with materials for self-promotion they should go in a separate repo, again we would be happy to link to from the docs.

Rgtemze replied:
Let me clear. You are suggesting that in Bokeh's web-site you will provide link to our web-page served by us containing this example instead of contributing to _examples/app_ and being served by you. In that way, our source code will be in another repository.

Rgtemze commented:
We have been endeavoring to find the root cause of that unicode character error; however, we cannot locate where it is. Is it possible to find at least in which file the error is raised?

bryevdv commented:
> We have been endeavoring to find the root cause of that unicode character error; however, we cannot locate where it is. Is it possible to find at least in which file the error is raised?

I have other tasks all morning but I will see if I can look this afternoon.

bryevdv commented on examples/app/bokeh-decision-tree/Decision_Tree/ID3_Decision_Tree/id3_decision_tree.py:
```diff
@@ -0,0 +1,389 @@
+    """ For instance, there is a attribute which has values like "low", "med", "high" """
+    """ There are 100 instances. For every instances, the number of instances distribute like [30, 40, 30]"""
+    """ For instance, label values are like "un_acc", "acc", "good", "v_good" """
+    """ Instances divide according to the label values like [[10, 5, 15, 0], [10, 10, 10, 10], [4, 13, 3, 10]]"""
```
This should be a single triple quoted doc string, not multiple strings

bryevdv commented on examples/app/bokeh-decision-tree/Decision_Tree/ID3_Decision_Tree/bucheim.py:
```diff
@@ -0,0 +1,159 @@
+    """
+        Calling FIRSTWALK(node) computes a preliminary x-coordinate for node. Before that, FIRSTWALK is
+        applied recursively to the children of node, as well as the function APPORTION. After spacing out the
+        children by calling EXECUTESHIFTS, the node  is placed to the midpoint of its outermost children.
```
Docstrings in Bokeh typically use triple single quotes, i.e. `'''` and docstring always starts on the same line as the first triple quotes. Ideally that line is a single short intro/decarative sentence, and then after a blank line, any additional details. We use google style parameter docs. Here is a prototypical example including:

```
def find(objs, selector, context=None):
    ''' Query a collection of Bokeh models and yield any that match a selector.

    Queries are specified as selectors similar to MongoDB style query
    selectors, as described for :func:`~bokeh.core.query.match`.

    Args:
        obj (Model) : object to test
        selector (JSON-like) : query selector
        context (dict) : kwargs to supply callable query attributes

    Returns:
        Model : objects that match the query
    
    '''
```

Rgtemze replied:
We will fix these as soon as possible.

Rgtemze commented:
@bryevdv Could the unicode error occur because of backslah(\) character or tab character (\t)?

Rgtemze commented:
@bryevdv We finally realized that unicode error was raised because of image files (i.e .jpg and .jpeg). We will publish another commit tomorrow or the day after tomorrow fixing all the problems.

bryevdv commented:
@Rgtemze I am glad you were able to figure it out. I am somewhat underwater with other work. But now that things are green I will review this PR more closely in the next few days.

bryevdv commented:
@Rgtemze OK I had a chance to check out the branch, and run the example. It looks quite nice I think it will be a good addition. Here are a few initial comments:

* app dir should be renamed `decision_tree`, all dir names in the repo use snake case throughout, and the "bokeh" part is superfluous

* The "arrow labels" look like the text is anchored on the left, I think you should be able to set a text property on the label to anchor it at the center of the text. I believe that would look slightly better as well as respond better during zooms

* The "axis-only" zoom should be turned off, it's confusing without visible axes: https://bokeh.pydata.org/en/latest/docs/reference/models/tools.html#bokeh.models.tools.WheelZoomTool.zoom_on_axis

* The arrow labels can overlap alot:

    <img width="569" alt="screen shot 2018-07-30 at 14 57 05" src="https://user-images.githubusercontent.com/1078448/43425918-dd7789d6-9408-11e8-934c-2af643423abb.png">

    If there is not otherwise a simple way  to prevent that, a simple measure might be to give those labels a slight rotation, so that they don't stack up on one another. The labels also seem a bit far from the nodes. 

* I would consider hiding the toolbar altogether 

I will any code review comments inline in the PR as I have time to look closer.

EvrimOzel commented:
Hi, I'm one of the contributors of this example. I've modified the example to include your suggestions, and I also removed the label text completely, seeing that giving an angle didn't look very good and it didn't solve the overlapping issue (users can already see a particular label by moving the cursor on leaf nodes).

bryevdv commented:
Hi, I've left some initial code-related review. Once you provide provenance/license info about the data sets I will upload the large file to S3 and provide you with a new module for `bokeh.sampledata` to access it, then it can be deleted from the PR.

Rgtemze commented on examples/app/decision_tree/src/data/mushrooms.txt:
```diff
@@ -0,0 +1,8125 @@
+cap-shape,cap-surface,cap-color,bruises,odor,gill-attachment,gill-spacing,gill-size,gill-color,stalk-shape,stalk-root,stalk-surface-above-ring,stalk-surface-below-ring,stalk-color-above-ring,stalk-color-below-ring,veil-type,veil-color,ring-number,ring-type,spore-print-color,population,habitat,class
```
It is mushroom data set from UCI Machine Learning Repository https://archive.ics.uci.edu/ml/datasets/mushroom.

bryevdv replied:
I see, I wish they stated directly that redistribution is permissible, but I will take their Citation policy to implicitly grant that, and make sure to include a citation in the sampledata module. I will upload the data files and and get you the data module by tomorrow.

Rgtemze replied:
Hi, @bryevdv Any update about the module?

Rgtemze commented on examples/app/decision_tree/src/data/lens.txt:
```diff
@@ -0,0 +1,25 @@
+Age,Spectacle Prescription,Astigmatic,Tear Production Rate,Classes
```
It is lens data set also from UCI, https://archive.ics.uci.edu/ml/datasets/lenses.

Rgtemze commented on examples/app/decision_tree/src/tree/bucheim.py:
```diff
@@ -0,0 +1,152 @@
+    '''
+    node.coord = (node.depth, node.prelim + m + 2)
+    for child in node.children:
+        second_walk(child, m + node.mod)
```
Could you clarify more what you meant?

bryevdv replied:
I mean only that you should make function definitions be in alphabetical order when possible, and that this ordering should be done for all the modules in the PR.

bryevdv commented:
Hi, I've gotten caught up in some other work that takes priority, I have uploaded the data to here:

https://s3.amazonaws.com/bokeh_data/mushrooms.txt

For the module, you should make a `mushrooms.py` in `bokeh/sampledata` that looks basically like the code below. Note that you can pass any additional pandas `read_csv` keyword args to `external_csv` if you need to. Please also fill in the module docstring with the data provenance information. 


```
#-----------------------------------------------------------------------------
# Copyright (c) 2012 - 2017, Anaconda, Inc. All rights reserved.
#
# Powered by the Bokeh Development Team.
#
# The full license is in the file LICENSE.txt, distributed with this software.
#-----------------------------------------------------------------------------
'''

'''

#-----------------------------------------------------------------------------
# Boilerplate
#-----------------------------------------------------------------------------
from __future__ import absolute_import, division, print_function, unicode_literals

import logging
log = logging.getLogger(__name__)

#-----------------------------------------------------------------------------
# Imports
#-----------------------------------------------------------------------------

# Standard library imports

# External imports

# Bokeh imports
from ..util.sampledata import external_csv

#-----------------------------------------------------------------------------
# Globals and constants
#-----------------------------------------------------------------------------

__all__ = (
    'data',
)

#-----------------------------------------------------------------------------
# General API
#-----------------------------------------------------------------------------

#-----------------------------------------------------------------------------
# Dev API
#-----------------------------------------------------------------------------

#-----------------------------------------------------------------------------
# Private API
#-----------------------------------------------------------------------------

#-----------------------------------------------------------------------------
# Code
#-----------------------------------------------------------------------------

data = external_csv('mushrooms', 'mushrooms.csv', sep=',')
```

Rgtemze commented:
Hi, @bryevdv. 
We created `mushrooms.py` and put it inside `bokeh/sampledata` folder but we could not use it since importing it with `from bokeh.sampledata.mushrooms import data as mushroom_data
` created an error:

> ModuleNotFoundError: No module named 'bokeh.sampledata.mushrooms'

bryevdv commented:
You will need to  install the dev version from source into your local python environment.

bryevdv commented:
You will have to install from *local source checkout*, i.e. by running `setup.py install` or `setup.py develop`, since that is where the changes are.  Any version you download and install will not have the file, so the error you are seeing is entirely expected.

Rgtemze commented:
Hi @bryevdv, we prepared the `mushrooms.py` and added it to `bokeh/sampledata`. Now, our code works well with `bokeh.sampledata.mushrooms`. We also updated `util/sampledata` so that `mushrooms.txt` could be fetched from S3.

bryevdv commented:
@Rgtemze great, I had overlooked the s3 line, thanks for catching it. Can you now remove the `mushrooms.txt` file that is in the branch? 

I'd like to do another round of review after this just to see if the code can be tightened up or organized any more clearly. This example has much more code than our other examples so we need to spend the effort or make sure it is as polished and consistent with other existing code a possible. Hopefully we can get @mattpap  to also send a little time reviewing. Just to set expectations, though: I have a conference to prepare for and attend next week, and @mattpap is heads-down in other other work over roughly the same time. So, it will probably be week after next before we can turn more attention to this PR. From my perspective there is no big hurry since the next release is still some time away, there is no need to rush. 

Offhand, if you wanted to tweak things in the mean time, I think it would be good to try to shorten `main.py` somewhat considerably. Generally it's good if the `main.py` is just used to "put the pieces together" which I would loosely say means that it should be calling functions, but not defining many functions (except perhaps bokeh callbacks).

Rgtemze commented:
@bryevdv We are looking forward to your reviews. We will try to edit the code as much as possible to fulfill the requirements of Bokeh. We wish you a good conference.

bryevdv commented:
Hi @Rgtemze @EvrimOzel I have gotten back from my conference however I am no inundated with other work. But I have had a chance to reflect a little on this PR and I would like to propose an alternate course of action. 

I've had a nagging concern about adding this example, and after some time away I can elucidate it clearly. Examples in the Bokeh repo should, first and foremost, demonstrate something specific about Bokeh, and secondarily to be a good pedagogical reference, should try to do so in a very focused way, which is where I get hung up on adding this example. It is over 1200 lines of code, and more than half of them have nothing to do with Bokeh at all. I don't think that fits in well with the existing examples, and I don't think it will be a helpful general reference for bokeh users, because of its sheer size.  

However, I do think the example itself is very impressive, and interesting, and a wonderful showcase of what's possible with Bokeh in a bigger context.  So that's exactly how I would propose to proceed, by having you put this code in your own GitHub repo, instead of Bokeh's and making a showcase page of "projects using Bokeh" in the docs that can link to it with screenshots (and we can also periodically blog/tweet about it as well). I think such a showcase is long overdue anyway, and I think this project is an idea seed to start it with. 

Please let me know your thoughts.

Rgtemze commented:
Hi @bryevdv, this offer sounds great, we will be glad to do it. We already have a website containing the tree visualiser, although it includes some features that are removed in this pull request like dataset uploading, train-test data splitting e.t.c. Actually, this visualiser is part of our project in which we aimed to visualise Machine Learning algorithms and interactively explain them. We also visualised the artificial neural network which we can show it to you. Besides, we already have the Github Repo for the tree visualizer.
We are looking forward to hearing you the next steps to go through.

bryevdv commented:
@Rgtemze great, I am going to go ahead and close this PR to keep the PR list tidy, but we can continue the discussion of details here. Basically I would ask fo these things:

* A URL to the GH project
* A URL to the live hosted app (if applicable)
* A screenshot or two that you would like to represent the project
* A few-sentences description (that I may tweak/edit some)

These things I will plan to put in a new page in the docs that lists "Projects using Bokeh" and I will commit to having this page included in the 1.0 release documentation (which will be in the next month or so)

Rgtemze commented:
- https://github.com/SBNC-Bavlab/bokeh-decision-tree
- http://makineogren.me/tree is where the tree resides. We bought the server from Digital Ocean but the server runs on a slow machine (1 GB ram and 1 CPU). Therefore, if multiple users interact with the tree, delay is inevitable. We will hopefully upgrade to a better server. As I mentioned before, this visualizer is the part of our project which is hosted on http://makineogren.me. We prepared it in Turkish, as in Turkey people suffer from lack of Turkish source in Machine Learning; however, we can readily translate it into English.
- ![Pic1](https://user-images.githubusercontent.com/8372277/44806627-22313e80-abd0-11e8-8569-53c8ad3b05e2.png)
![Pic2](https://user-images.githubusercontent.com/8372277/44806629-22313e80-abd0-11e8-939e-fb5859cc5f52.png)
- "This is a decision tree visualizer which you can interact with by selecting the dataset, choosing the root attribute of the decision tree, toggling the attributes and adjusting the train-test split percentage."

Rgtemze commented:
Hi @bryevdv, you wanted a few-sentences description. Is it about the project or the tree itself? I apparently wrote a description about the tree in the above post. However, I can provide you with the description of our project itself.

Rgtemze commented:
@bryevdv Do you still consider referring to our Decision Tree Visualizer?

bryevdv commented:
@Rgtemze I do, I am just  currently overtasked and have more things to do than I can accomplish. I have  four projects lined up for the showcase page so it's definitely something I want to do. But I am taking all of November off from OSS development, so a this stage it will probably not be until december.

Rgtemze commented:
Hi @bryevdv , I just wanted to remind you of adding our project to the showcase page. Would you like us to do any preparation for it?

bryevdv commented:
@Rgtemze for this I would like to start up a new subdomain `stories.bokeh.org` analogous to http://stories.dask.org/en/latest/ for the showcase. However, the current situation is that funded development around Bokeh is at a low point (i.e. in the  past more people were paid purely to spend work time developing bokeh, now much less) Given this reality, and available resources and priorities, I don't know when it will be possible to stand the site up at this point, unfortunately.

Rgtemze commented:
@bryevdv I am very thankful to you for informing us. It is not a problem at all, we don't have a rush. I wish you a good luck in your works.
